modules/foreman_facts.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# (c) 2018, Sean O'Keeffe <seanokeeffe797@gmail.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: foreman_facts
+short_description: Gather facts about Foreman Entities
+description:
+  - "Gather facts about Foreman Entites"
+author:
+  - "Sean O'Keeffe (@sean797)"
+requirements:
+  - nailgun
+options:
+  server_url:
+    description:
+      - URL of Foreman server
+    required: true
+  username:
+    description:
+      - Username on Foreman server
+    required: true
+  password:
+    description:
+      - Password for user accessing Foreman server
+    required: true
+  verify_ssl:
+    description:
+      - Verify SSL of the Foreman server
+    default: true
+    type: bool
+  entity:
+    description:
+      - Entity to search
+      - Run with parameter set to C(foo) to see all available options.
```

Comment:
I am not sure, i like this approach. It would feel more natural to me if a missing entity option lead to the list, and an invalid entity rose an error. Writing stable playbooks would be easier.

Reply:
The error is raised by Ansible based on 

```python

choices=nailgun_entites(), required=True

```

If it's missing you get `"msg": "missing required arguments: entity"` if you do, but its not valid you get `msg": "value of entity must be one of: ActivationKey, ..... `

I don't think its worth us maintaining our own error handling code for this. If you think a missing option error should include a list or possible options, then it may be worth raise that directly with the Ansible community.

Reply:
So you are saying, an inexistent entity raises an error with a suitable message. I still find it a little bit strange to advertise raising an error without saying so in the docs part.
Do you mind writing something like 'For an invalid entity like foo the error message contains a list if valid ones.'?

Reply:
Good idea, updated! Thanks :+1: